homeassistant/components/watson_iot.py

```diff
@@ -0,0 +1,215 @@
+"""
+
+import logging
+import queue
+import threading
+import time
+
+import voluptuous as vol
+
+from homeassistant.const import (
+    CONF_DOMAINS, CONF_ENTITIES, CONF_EXCLUDE, CONF_INCLUDE,
+    CONF_TOKEN, CONF_TYPE, EVENT_STATE_CHANGED, EVENT_HOMEASSISTANT_STOP,
+    STATE_UNAVAILABLE, STATE_UNKNOWN)
+from homeassistant.helpers import state as state_helper
+import homeassistant.helpers.config_validation as cv
+
+REQUIREMENTS = ['ibmiotf==0.3.4']
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_ORG = 'organization'
+CONF_ID = 'id'
+
+DOMAIN = 'watson_iot'
+
+RETRY_DELAY = 20
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.All(vol.Schema({
+        vol.Required(CONF_ORG): cv.string,
+        vol.Required(CONF_TYPE): cv.string,
+        vol.Required(CONF_ID): cv.string,
+        vol.Required(CONF_TOKEN): cv.string,
+        vol.Optional(CONF_EXCLUDE, default={}): vol.Schema({
+            vol.Optional(CONF_ENTITIES, default=[]): cv.entity_ids,
+            vol.Optional(CONF_DOMAINS, default=[]):
+                vol.All(cv.ensure_list, [cv.string])
+        }),
+        vol.Optional(CONF_INCLUDE, default={}): vol.Schema({
+            vol.Optional(CONF_ENTITIES, default=[]): cv.entity_ids,
+            vol.Optional(CONF_DOMAINS, default=[]):
+                vol.All(cv.ensure_list, [cv.string])
+        }),
+    })),
+}, extra=vol.ALLOW_EXTRA)
+
+
+def setup(hass, config):
+    """Set up the Watson IoT Platform component."""
+    from ibmiotf import gateway
+
+    conf = config[DOMAIN]
+
+    include = conf.get(CONF_INCLUDE, {})
+    exclude = conf.get(CONF_EXCLUDE, {})
+    whitelist_e = set(include.get(CONF_ENTITIES, []))
+    whitelist_d = set(include.get(CONF_DOMAINS, []))
+    blacklist_e = set(exclude.get(CONF_ENTITIES, []))
+    blacklist_d = set(exclude.get(CONF_DOMAINS, []))
+
+    client_args = {
+        'org': conf.get(CONF_ORG),
+        'type': conf.get(CONF_TYPE),
+        'id': conf.get(CONF_ID),
+        'auth-method': 'token',
+        'auth-token': conf.get(CONF_TOKEN),
+    }
+    watson_gateway = gateway.Client(client_args)
+
+    def event_to_json(event):
+        """Add an event to the outgoing list."""
+        state = event.data.get('new_state')
+        if state is None or state.state in (
+                STATE_UNKNOWN, '', STATE_UNAVAILABLE) or \
+                state.entity_id in blacklist_e or state.domain in blacklist_d:
+            return
+
+        try:
+            if (whitelist_e and state.entity_id not in whitelist_e) or \
+                    (whitelist_d and state.domain not in whitelist_d):
+                return
+
+            _include_state = _include_value = False
+
+            _state_as_value = float(state.state)
+            _include_value = True
+        except ValueError:
+            try:
+                _state_as_value = float(state_helper.state_as_number(state))
+                _include_state = _include_value = True
+            except ValueError:
+                _include_state = True
```

Comment:
False ?

Reply:
You know I'm looking at this again, and I'm not sure what my intent was when I wrote this a month ago. False wouldn't make much sense because if we get a ValueError there _include_state will already be false. I can only assume I meant for it to be true, but I can't remember why. I'm just going to leave it as is, and if I remember what my intent was I'll add a comment here.

Reply:
Well I think that using exception handling to implement logic is not the best approach... maybe you can rework this part of the code into something more readable and less prone to confusion.